Boost: Cache Preloading Without Force Cache Invalidation

LiamSarsfield opened the pull request:
> [!NOTE]
> Merge this PR after https://github.com/Automattic/jetpack/pull/42453 is merged.

## Fixes
#42111

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Adds a Cache Preloading system for Cornerstone Pages
* Introduces a new submodule `Cache_Preload.php` to handle cache preloading logic (Submodule of Page Cache)
* Updates cache invalidation process to trigger preloading for cornerstone pages when they're invalidated
* Improves the structure of the `invalidate` method in `File_Storage.php` to include action handling

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
pc9hqz-3nV-p2

## Does this pull request change what data or activity we track or use?
N/A

## Testing instructions:

### Prerequisites
* Ensure that the constant `JETPACK_BOOST_ALPHA_FEATURES` is defined as true either in an mu-plugin or wp-config.php.
* If testing on a Jurassic Ninja site, ensure Edge Caching is disabled in Settings -> Edge Cache -> Disable Edge Cache
  * Bypassing Edge Cache is out of scope for this PR.

### 1. Saving Cornerstone Pages list, Clearing Page Cache
<details>
<summary>Instructions</summary>

#### Before Applying Branch
1. Go to Boost Settings page.
2. Ensure Boost Premium is on the WordPress site.
3. Enable Optimize CSS Loading (Cloud CSS).
4. Enable Cache Site Pages (Page Cache).
5. Add at least 2 valid Cornerstone Pages.
6. Clear the Page Cache through the wp-admin UI
7. Within the WordPress site, head to `./wp-content/boost-cache/cache/(domain)`
8. Verify that **no** page caching files are present.
9. Head to one of the valid Cornerstone Pages added in Incognito.
10. Verify that page caching is generated for that Cornerstone Page visited.
11. Clear Page Cache.
12. Within the WordPress site, head to `./wp-content/boost-cache/cache/(domain)`
13. Verify that **no** page caching files are present.

#### After Applying Branch
1. Go to Boost Settings page.
2. Add a third valid page to the Cornerstone Pages list
3. Refresh the page (to ensure Preloading occurs)
4. Within the WordPress site, head to `./wp-content/boost-cache/cache/(domain)`
5. You should now see that 3 caching files have been generated automagically ✨
6. Clear the Page Cache through the wp-admin UI
7. Refresh the page (to ensure Preloading occurs within the Cron)
5. You should now see that 3 caching files have been generated automagically again ✨
</details>

### 2. Updating Cornerstone Page
<details>
<summary>Instructions</summary>

#### Before Applying Branch

1. Go to Boost Settings page.
2. Ensure Boost Premium is on the WordPress site.
3. Enable Optimize CSS Loading (Cloud CSS).
4. Enable Cache Site Pages (Page Cache).
5. Add at least 2 valid Cornerstone Pages.
6. Clear the Page Cache through the wp-admin UI
7. Within the WordPress site, head to `./wp-content/boost-cache/cache/(domain)`
8. Verify that **no** page caching files are present.
9. Edit one of the posts/pages you added to the Cornerstone Page list.
10. Within the WordPress site, head to `./wp-content/boost-cache/cache/(domain)`
11. Verify that the page you edited is not cached.

**After Applying Branch**
1. Go to Boost Settings page.
2. Clear Page Cache by deleting the `./wp-content/boost-cache/cache/(domain)` folder
3. Edit one of the posts/pages you added to the Cornerstone Page list.
4. Refresh the post edit page (to ensure Preloading occurs within the Cron).
5. Within the WordPress site, head to `./wp-content/boost-cache/cache/(domain)`
6. You should now see that specific page's caching file generated automagically ✨
</details>

### 3. Feature Flag Testing
<details>
<summary>Instructions</summary>

1. Remove the constant `JETPACK_BOOST_ALPHA_FEATURES` added previously, or set it to false.
2. Redo the testing for Saving Cornerstone Pages list, Clearing Page Cache, and Updating Cornerstone Page. Specifically the "Before Applying Branch" should now the expected behaviour, even after applying the Branch.
</details>

LiamSarsfield committed:
Page Cache: Add support for submodules with Cache Preload

Co-authored-by: Peter Petrov <dilirity@users.noreply.github.com>

LiamSarsfield committed:
changelog

LiamSarsfield committed:
Page Cache: Enhance Cache Preload with dynamic post scheduling and update mechanism

- Refactor preload mechanism to support dynamic post scheduling
- Add method to handle post updates and schedule preloading for cornerstone pages
- Implement more flexible preload post tracking using WordPress options
- Improve error handling during page preloading

LiamSarsfield committed:
Boost: Enhance Cache Preloading and Cornerstone Page Detection

- Add new method `is_cornerstone_page_by_url()` in Cornerstone_Utils to check cornerstone pages by URL
- Modify `is_cornerstone_page()` to use the new URL-based method
- Update Cache_Preload to handle cache invalidation and preloading of cornerstone pages
- Add action hook for successful cache invalidation
- Improve method names and add comments for clarity

LiamSarsfield committed:
Boost: Improve Cache Preload Handling for Cornerstone Pages

- Enhance `handle_cache_invalidation()` method to support different cache invalidation types
- Add support for full cache deletion by preloading all Cornerstone Pages
- Improve documentation with detailed comments and @since tags
- Update method signature to include invalidation type parameter
- Refactor cache preloading logic for more robust handling of cache events

LiamSarsfield commented on projects/plugins/boost/app/modules/optimizations/page-cache/Cache_Preload.php:
```diff
@@ -0,0 +1,231 @@
+		// Clear the preload posts so they're not preloaded again.
+		$this->set_posts_to_preload( array() );
+
+		foreach ( $posts as $post ) {
```
I wonder if we should limit how many pages/posts are being preloaded here within the cron to prevent potentially overloading the site. Theoritically the limit of URLs should only be 10 as that's our current Cornerstone Page limit.
We could only preload 10 URLs at a time, and then schedule another cron job after 2 seconds to repeat the same.
Not sure if it's worth doing now, or wait until Version 2.

LiamSarsfield replied:
I implemented this now with batching, as seen below:
https://github.com/Automattic/jetpack/blob/cc070a260bfc53c86bc7dd7d2c2ce784942b24c3/projects/plugins/boost/app/modules/optimizations/page-cache/Cache_Preload.php#L145-L172

LiamSarsfield committed:
Boost: Improve Cache Preload Performance and Error Handling

- Refactor page preloading to process URLs in batches of 10
- Add batch processing to reduce server load and improve reliability
- Enhance error handling with detailed logging for preload failures
- Improve HTTP response validation during page preloading
- Add methods to manage batch processing and track remaining URLs

LiamSarsfield committed:
Boost: Add Comprehensive Unit Tests for Cache Preload Module

- Create new test suite for Cache_Preload class in Jetpack Boost
- Implement comprehensive test coverage for cache preloading methods
- Test scenarios including cornerstone page preloading, cache invalidation, and scheduling
- Verify interactions with WordPress core functions and utility classes
- Ensure robust testing of preload scheduling, post update handling, and cache management

LiamSarsfield committed:
Boost: Update Cache Preload Test Suite with Direct Method Calls

- Refactor test cases to use direct method calls instead of partial mocks
- Replace Mockery-based mocking with explicit WordPress function expectations
- Improve test coverage by directly testing Cache_Preload class methods
- Simplify test setup and reduce complexity of test assertions
- Ensure consistent testing approach for cache preloading functionality

LiamSarsfield committed:
Boost: Add Cache-Busting Headers to Page Preload Requests

- Enhance page preloading by adding cache-control headers

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/Cache_Preload.php:
```diff
@@ -0,0 +1,307 @@
+	 * @since $$next-version$$
+	 * @return void
+	 */
+	public function setup() {
```
Small note. I like the comments, however they would be the same for all modules. We can use the following to get the comments:

```php
public function setup() { // @inheritDoc
```

However, I don't know if that works with `@since`.

LiamSarsfield replied:
Per our discussion, this (and other functions from interfaces) has been updated now. It does look like it is aware  enough to prepend the existing doc from the interface per attached.
![Screenshot 2025-03-12 at 08 53 07](https://github.com/user-attachments/assets/d819edd2-0ead-47a1-a634-288f9d1f9c33)

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/Cache_Preload.php:
```diff
@@ -0,0 +1,307 @@
+	 */
+	public function schedule_preload_cronjob() {
+		if ( ! wp_next_scheduled( 'jetpack_boost_preload_pages' ) ) {
+			// Add 2 seconds delay to prevent potential race conditions with the cronjob, or so the cache invalidation process has time to complete.
```
Which cronjob are we referring to? Is it possible for the invalidation process to take more than 2 seconds?

Just wondering if we need the `+ 2` part.

LiamSarsfield replied:
I noticed that for Cloud CSS specifically the cache may be invalidated and preloaded twice in quick succession without a delay by doing:

1. Ensure Cloud CSS/Page Cache is enabled.
2. Invalidate Cache.
3. Update Cornerstone pages.

- The Preload triggered by Update Cornerstone pages would be done immediately in the cronjob.
- Cloud CSS would return the CSS, and hence invalidate the cache again.
- And hence would be rebuilt again from another cache preloading cron.

The 2 second timer is more of an optimization to prevent rapid cache rebuilds in the case that we attempt to schedule the cron job and one was already scheduled within 2 seconds, the cron job will not be scheduled twice within 2 seconds. I've updated the comment to clarify this as the original was misleading.
Let me know though if you think this is too much of a pre-optimization to be doing for an edge case.

LiamSarsfield replied:
Resolving this thread if that's okay as per offline discussion this timer is no longer needed as
the advantages of preloading for Cornerstone Pages immediately upon the list being updated, outweigh the disadvantages of relying on the Cloud CSS invalidating the cache, which would preload the Cornerstone pages.

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/Cache_Preload.php:
```diff
@@ -0,0 +1,307 @@
+	public function schedule_preload( $post_to_schedule ) {
+		$posts = $this->get_posts_to_preload();
+		if ( is_array( $post_to_schedule ) ) {
+			$posts = array_merge( $posts, $post_to_schedule );
```
`array_unique` would help eliminate any duplicates we might have after the merge.

I saw this is done in `set_posts_to_preload` but I think it should be done here. Otherwise you'd have to make sure the array is okay before using it.

LiamSarsfield replied:
True, I think it shouldn't be the responsibility of the setter alone for this. 
I've updated the `add_to_queue` and `prepare_next_batch` to also call `array_unique` to ensure the remaining posts validated early, and often.

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/Cache_Preload.php:
```diff
@@ -0,0 +1,307 @@
+	 * @param array $posts Array of post URLs to preload.
+	 * @return void
+	 */
+	public function set_posts_to_preload( array $posts ) {
```
This is the only method that has a typed parameter.

LiamSarsfield replied:
Good catch, I've updated all functions to ensure they have typed parameters now, except for `schedule_preload` which accepts either a string or an array. And only PHP 8.0 supports [union types](https://php.watch/versions/8.0/union-types).

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/Cache_Preload.php:
```diff
@@ -0,0 +1,307 @@
+	 */
+	private function prepare_next_batch( $posts ) {
+		// Process in batches of 10 to reduce server load
+		$batches       = array_chunk( $posts, 10 );
```
Let's move this to the top so we can easily configure it.

LiamSarsfield replied:
I originally though you meant the function 🤦 but then when revisiting I believe you meant the batch size.
I've added this as a constructor argument within the new `Cache_Preload_Queue_Manager` class to make it more obvious that it can be configurable if need be.

dilirity replied:
Oh shoot. Sorry, should've been clear. I meant the size.

I think what you did folds nicely in https://github.com/Automattic/jetpack/pull/41988#discussion_r1988842585 😅

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/Cache_Preload.php:
```diff
@@ -0,0 +1,307 @@
+	private function process_batch( $batch ) {
+		foreach ( $batch as $url ) {
+			try {
+				$this->preload_page( $url );
```
I'm not sure we need the try-catch since nothing throws an exception in `preload_page`.

LiamSarsfield replied:
Good point. I've removed it, especially as `remote_get` does not throw an exception but instead returns a `WP_Error`

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/Cache_Preload.php:
```diff
@@ -0,0 +1,307 @@
+<?php
```
I know this is a work in progress and I'm not sure if you have plans on updating, but I just wanted to note it.

I feel like the `Cache_Preload` class is doing a bit more than it should. At the same time, we have other modules that do more than they should (just look at Critical CSS). I'm not sure how more complex the code would get if we move the preloading as a whole to a separate class.

I'd like to know your thoughts on this!

LiamSarsfield replied:
Good point! From revisiting this again I agree it was doing quite a lot. I've created a new class called `Cache_Preload_Queue_Manager` that extracts the queue management logic that is more self-contained. I could break it down further by having another class that handles the HTTP requests, but also want to balance complexity level and cohesion of the functionality.

dilirity replied:
I think keeping it at two levels deep is a good balance. The main class and then helper classes to handle implementation details.

dilirity reviewed (commented):
Great work on this!

I have a few suggestions

LiamSarsfield committed:
Boost: Refactor Cache Preload Module with Feature Flag and Improvements

- Add feature flag support for cache preloading via JETPACK_BOOST_ALPHA_FEATURES constant
- Improve type hinting and method signatures in Cache_Preload class
- Update test suite to support new feature flag implementation

LiamSarsfield committed:
Boost: Implement Always-On Cache Preload and Improve Scheduling Logic

- Add `Is_Always_On` interface to Cache_Preload class
- Enhance cron job scheduling comment to explain 2-second delay rationale

LiamSarsfield committed:
Boost: Add Cache Preload Queue Management Class

- Implement `Cache_Preload_Queue_Manager` to handle preload queue operations
- Add methods for managing, adding, and processing preload URL batches
- Provide queue management functionality with clear, modular design

LiamSarsfield committed:
Boost: Optimize Cache Preload Queue Management

- Add in-memory caching for preload queue to reduce database calls
- Improve `get_posts_to_preload()` and `set_posts_to_preload()` methods
- Refactor `add_to_queue()` to use the new in-memory cache mechanism
- Simplify queue management with a single source of truth

LiamSarsfield committed:
Boost: Refactor Cache Preload Queue Manager with Batch Size Configuration

- Add constructor to set configurable batch size for preloading
- Update `prepare_next_batch()` method to use instance batch size
- Remove redundant batch size parameter from method signature

LiamSarsfield committed:
Enhance Cache Preload Documentation and Logic

- Update comments in Cache Preload class to clarify cache management processes
- Refactor conditional checks in File Storage class for improved readability and logic flow
- Ensure accurate descriptions of cache invalidation and rebuilding processes

LiamSarsfield committed:
Remove Cache Preload Queue Manager and Refactor Cache Preload Logic

- Delete Cache_Preload_Queue_Manager class to simplify preload management
- Refactor Cache_Preload class to directly handle preload queue operations
- Clear preload queue after processing to ensure accurate state management

LiamSarsfield committed:
Update Cache Preload Scheduling Logic to Remove Delay

- Adjusted the scheduling of the preload cron job to occur immediately instead of with a 2-second delay

LiamSarsfield committed:
Refactor Cache Preload Logic to Use Direct Page Parameter

- Removed unnecessary variable assignment for the page URL in the preload_page method

dilirity committed:
Fix My jetpack assets enqueued on every page instead of just Boost's

dilirity committed:
add changelog

haqadn commented on projects/plugins/boost/app/modules/optimizations/page-cache/class-cache-preload.php:
```diff
@@ -0,0 +1,202 @@
+		if ( ! wp_next_scheduled( 'jetpack_boost_preload_pages' ) ) {
+			wp_schedule_single_event( time(), 'jetpack_boost_preload_pages' );
+		}
+	}
```
Could we have created a cron job that takes the list of URLs as arguments?
So it would look like `wp_schedule_single_event( time(), 'jetpack_boost_preload_pages', [ ...$urls ] );`

That might help us avoid worrying about pulling/merging on the one array for all preloads.

WordPress [automattically deduplicates](https://developer.wordpress.org/reference/functions/wp_schedule_single_event/#:~:text=time%20has%20passed.-,Note%20that%20scheduling%20an%20event%20to%20occur%20within%2010%20minutes%20of,values%20for%20each%20scheduled%20event.,-Use%20wp_next_scheduled()) the same cron job with identical arguments if they are scheduled within a  10 minutes window. It appears, most of the time, we would be preloading individual pages. So passing the URLs as aruments seem much simpler to me. What do you think?

LiamSarsfield replied:
Good to know about the automate deduplications. That should help if the event is potentially scheduled multiple times rapidly. I've updated the PR to reflect your suggestion.

LiamSarsfield committed:
Merge remote-tracking branch 'origin/trunk' into add/boost/cache-preloading-cornerstone-pages

LiamSarsfield committed:
Refactor Cache Preload Methods to Accept Post URLs as Parameters

- Modified preload_pages to accept an array of posts as a parameter.
- Removed the old schedule_preload method and its associated logic.
- Adjusted handle_post_update to directly schedule preload for cornerstone pages.

LiamSarsfield committed:
Update Cache Preload to Use Array Parameters for Scheduling

- Refactored schedule_preload_cronjob to accept an array of posts.
- Adjusted tests to reflect changes in method signatures and ensure proper scheduling behavior.

LiamSarsfield committed:
Update test method names in Cache_Preload_Test to reflect cronjob functionality

- Renamed test methods to clarify that they are testing the schedule_preload_cronjob functionality.

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/class-cache-preload.php:
```diff
@@ -0,0 +1,173 @@
+	 * Schedules the preload cronjob, if not already scheduled to execute within 10 minutes with the same arguments, per wp_schedule_single_event.
+	 *
+	 * Sets up a single event to trigger the preload process with a short delay of 2 seconds
+	 * to prevent multiple rapid cache rebuilds when multiple events trigger in sequence.
```
The comment is outdated and can be move to multiple lines.

```suggestion
	 * Schedules the preload cronjob, if not already scheduled
	 * to execute within 10 minutes with the same arguments, per wp_schedule_single_event.
```

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/class-cache-preload.php:
```diff
@@ -0,0 +1,173 @@
+	 * Preloads the pages scheduled for preload.
+	 *
+	 * This method is called via a cronjob and processes pages in batches
+	 * to populate the cache.
```
```suggestion
	 * Preloads the pages scheduled for preload.
```

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/class-cache-preload.php:
```diff
@@ -0,0 +1,173 @@
+	 * Handle post updates to check if the post is a cornerstone page and schedule preload if needed.
+	 *
+	 * Triggered when a post is updated. If the post is identified as a cornerstone page,
+	 * its cache will be preloaded.
```
I feel like this is mentioning the same thing twice.

dilirity reviewed (changes requested):
Nice work! Love the testing instructions.

A few small things regarding the comments below.

Also - I noticed that if you don't have cache enabled and no cache is generated, when you enable caching, cornerstone pages aren't preloaded. Is this intentional?

Another thing I noticed is that when I remove a cornerstone page from the list, it still gets preloaded.

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/class-cache-preload.php:
```diff
@@ -0,0 +1,173 @@
+			return \JETPACK_BOOST_ALPHA_FEATURES === true;
+		}
+
+		return true;
```
I missed this. It should be false since we haven't released the feature.

LiamSarsfield replied:
Great catch, I am not sure how I missed this neither.

LiamSarsfield committed:
Update Cache_Preload class to modify cronjob scheduling logic and streamline documentation

- Changed the return value of Cache Preloading feature check to false if not having features flag.
- Simplified comments for scheduling and preloading methods by removing unnecessary details.

LiamSarsfield commented:
@dilirity 
> Also - I noticed that if you don't have cache enabled and no cache is generated, when you enable caching, cornerstone pages aren't preloaded. Is this intentional?

Yep, it's intentional this isn't in this specific PR. A seperate PR will include this.

> Another thing I noticed is that when I remove a cornerstone page from the list, it still gets preloaded.

I'm unable to reproduce this locally (see below), but I'll spin up a JN site to confirm.

https://github.com/user-attachments/assets/934e2b9d-b137-4d6d-b99a-13a4ada4cdb1

After spinning up a JN site, I'm also unable to reproduce, am I reproducing it incorrectly?

https://github.com/user-attachments/assets/0e6f0cd6-b2c1-4061-a65a-6fb664260209

dilirity commented:
> After spinning up a JN site, I'm also unable to reproduce, am I reproducing it incorrectly?

Don't delete the cache folder via the console (the user wouldn't be doing that).

- Add two pages to the Cornerstone Pages list;
- Make sure there's cache for both;
- Remove one from the list and check the cache - there should be two pages.

dilirity commented:
> But do we clear the cache on the cornerstone update, @dilirity? If a cornerstone page becomes non-cornerstone, it should be cached until a GC has run. What am I missing?

Should it be cached? I wasn't sure. That's why I mentioned it. @haqadn

dilirity commented:
@haqadn helped me align with how preloading works. @LiamSarsfield everything works as expected.

dilirity reviewed (approved):
@haqadn helped me align with how preloading works. @LiamSarsfield everything works as expected.

:shipit:
